test/performance/odp_ipsec.c

```diff
@@ -654,38 +668,62 @@ run_measure_one(ipsec_args_t *cargs,
 
 	fill_time_record(start);
 
-	while ((packets_sent < cargs->iteration_count) ||
-	       (packets_received < cargs->iteration_count)) {
-		if ((packets_sent < cargs->iteration_count) &&
-		    (packets_sent - packets_received <
-		     cargs->in_flight)) {
-			odp_packet_t out_pkt;
-			int num_out = 1;
+	while ((packets_sent < cargs->packet_count) ||
+	       (packets_received < cargs->packet_count)) {
+		num_pkts = cargs->packet_count - packets_sent;
+
+		/* Enqueue up to burst size */
+		num_pkts = num_pkts > burst_size ? burst_size : num_pkts;
+
+		/* Enqueue up to (max in flight - current in flight) */
+		in_flight = packets_sent - packets_received;
+		pkts_allowed = max_in_flight - in_flight;
 
-			pkt = make_packet(pkt_pool, payload_length);
-			if (ODP_PACKET_INVALID == pkt)
+		/* Enqueue either a burst of packets or skip */
+		num_pkts = num_pkts > pkts_allowed ? 0 : num_pkts;
+
+		if (odp_likely(num_pkts)) {
+			odp_packet_t out_pkt[num_pkts];
+			odp_packet_t pkt[num_pkts];
+			int i;
+
+			if (odp_unlikely(make_packet_multi(pkt_pool,
+							   payload_length,
+							   pkt,
+							   num_pkts)))
 				return -1;
 
-			if (cargs->debug_packets)
-				odp_packet_print_data(pkt, 0,
-						      odp_packet_len(pkt));
+			if (cargs->debug_packets) {
+				for (i = 0; i < num_pkts; i++)
+					odp_packet_print_data(pkt[i], 0,
+							      odp_packet_len(pkt[i]));
+			}
+
+			num_out = num_pkts;
 
-			rc = odp_ipsec_out(&pkt, 1,
-					   &out_pkt, &num_out,
+			rc = odp_ipsec_out(pkt, num_pkts,
+					   out_pkt, &num_out,
 					   &param);
-			if (rc <= 0) {
-				ODPH_ERR("failed odp_ipsec_out: rc = %d\n", rc);
-				odp_packet_free(pkt);
+			if (odp_unlikely(rc <= 0)) {
+				ODPH_ERR("Failed odp_ipsec_out: rc = %d\n", rc);
+				odp_packet_free_sp(pkt, num_pkts);
 				break;
 			}
-			check_ipsec_result(out_pkt);
+
+			for (i = 0; i < num_out; i++)
+				check_ipsec_result(out_pkt[i]);
 
 			packets_sent += rc;
 			packets_received += num_out;
-			if (cargs->debug_packets)
-				odp_packet_print_data(out_pkt, 0,
-						      odp_packet_len(out_pkt));
-			odp_packet_free(out_pkt);
+			if (cargs->debug_packets) {
+				for (i = 0; i < num_out; i++)
+					odp_packet_print_data(out_pkt[i], 0,
+							      odp_packet_len(out_pkt[i]));
+			}
+
+			if (odp_unlikely(num_out != num_pkts))
+				odp_packet_free_sp(&pkt[num_out], num_pkts - num_out);
```

Comment:
This code makes the assumption that the number of consumed packets (rc) is the same as the number of output packets (num_out), which won't be true as soon as fragmentation offload gets enabled. It would be more robust to check if the number of consumed packets (i.e. rc) is not the same as the number of input packets and also use the number of consumed packet instead of the number of output packets in the freeing statement. The number of output packets is not directly relevant to freeing the unconsumed packets.

Reply:
@JannePeltonen So I'll replace num_out with rc in this code snippet and that should be good enough, right?

Reply:
Yes, that should fix it. There is already that kind of code in the async case, so this just becomes more similar.